src/main/src/main/java/org/geoserver/catalog/impl/AbstractCatalogFacade.java

```diff
@@ -105,4 +113,88 @@ protected void setId(Object o) {
             OwsUtils.set(o, "id", o.getClass().getSimpleName() + "-" + uid);
         }
     }
+
+    @SuppressWarnings("unchecked")
+    protected static <T extends CatalogInfo> T copy(T obj) {
+        ByteArrayOutputStream os = new ByteArrayOutputStream();
+        try (ObjectOutputStream oos = new ObjectOutputStream(os)) {
+            oos.writeObject(obj);
+            try (ObjectInputStream ois = new ObjectInputStream(new ByteArrayInputStream(os.toByteArray()))) {
+                return (T) ois.readObject();
+            }
+        } catch (IOException | ClassNotFoundException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    protected static <T extends CatalogInfo> T syncInfo(T obj, boolean deep, Catalog catalog) {
+        obj = unwrap(obj);
+        if (deep) {
+            obj = copy(obj);
+        }
+        if (catalog != null) {
+            OwsUtils.set(obj, "catalog", catalog);
+        }
+        return obj;
+    }
+
+    /**
+     * Most basic way to sync catalog facade.
+     *
+     * @param source the source catalog facade
+     * @param target the target catalog facade
+     * @param deep whether deep copies should be made out of the catalog info objects. This might be necessary for some
+     *     catalog facades to prevent it from resetting the catalog reference back to the original catalog.
+     */
+    public static void syncFromTo(CatalogFacade source, CatalogFacade target, boolean deep) {
+        target = ProxyUtils.unwrap(target, LockingCatalogFacade.class);
+        for (WorkspaceInfo w : source.getWorkspaces()) {
+            target.add(syncInfo(w, deep, null));
+        }
+
+        for (NamespaceInfo ns : source.getNamespaces()) {
+            target.add(syncInfo(ns, deep, null));
+        }
+
+        for (StoreInfo s : source.getStores(StoreInfo.class)) {
+            target.add(syncInfo(s, deep, target.getCatalog()));
+        }
+
+        for (ResourceInfo r : source.getResources(ResourceInfo.class)) {
+            target.add(syncInfo(r, deep, target.getCatalog()));
+        }
+
+        for (StyleInfo s : source.getStyles()) {
+            target.add(syncInfo(s, deep, target.getCatalog()));
+        }
+
+        for (LayerInfo l : source.getLayers()) {
+            target.add(syncInfo(l, deep, null));
+        }
+
+        for (LayerGroupInfo lg : source.getLayerGroups()) {
+            target.add(syncInfo(lg, deep, null));
+        }
+
+        for (MapInfo m : source.getMaps()) {
+            target.add(syncInfo(m, deep, null));
+        }
```

Comment:
This is unacceptable IMHO
- I don't think a community module should have this big of an impact in core without discussion
- This is replacing `facade.add(info)` by `facade.add(deserialize(serialize(info)))`, which is going to be catastrophic in terms of performance
- It seems to me there might be some implementation detail in `jdbconfig` that might require this, or for which this would be a work-around, but definitely can't go in core.
- The rationale is not clear nor explained

Reply:
@groldan

1. there is no impact on core behaviour, because the serialize/deserialize only happens if deep=true, but the default catalogue passes deep=false, so that the original behaviour is entirely preserved.

2. Regarding the rationale, the only thing that is a work-around to me is that I have to call serialize/deserialize for a deep copy because no other cloning method seems to work. I think it is inherently wrong that the catalog info object are reference copied instead of deeply copied. From one catalog to another. The result is that two catalog refer to the same object, while the object has a reference back to only one of the catalogs. That IMHO is dangerous and badly designed. Nonetheless , I didn't touch this, I only avoided code duplication.